### PR TITLE
HPCC-18963 Avoid critical sections in makeVoidType()/makeNullType() etc.

### PR DESCRIPTION
**Pull request by jpmcmu**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [X] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [X] My code follows the code style of this project.
  - [X] My code does not create any new warnings from compiler, build system, or lint.
- [X] The commit message is properly formatted and free of typos.
  - [X] The commit message title makes sense in a changelog, by itself.
  - [X] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [X] I have read the CONTRIBUTORS document.
- [X] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [X] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: common/deftype/deftype.cpp [hunk 1]
```diff
@@ -47,31 +47,39 @@ static IAtom * emptyAtom;
...
 static CStringTypeToTypeMap * datatt;
 static CStringTypeToTypeMap * vstt;
 static CStringTypeToTypeMap * qstt;
 static CUnicodeTypeToTypeMap * utt;
 static CUnicodeTypeToTypeMap * vutt;
 static CUnicodeTypeToTypeMap * u8tt;
-static CIntTypeInfo *itt[2][8];
-static CSwapIntTypeInfo *sitt[2][8];
+
+// Note: C++ standard does not guarantee std::atomic will initialize memory
```
Comment (ghalliday) [inline common/deftype/deftype.cpp:69]: This comment is not true for static variables - they are default zero-initialised.  The initialiser list is not required.

File: common/deftype/deftype.cpp [hunk 2]
```diff
@@ -1427,9 +1435,29 @@ bool CModifierTypeInfo::equals(const CTypeInfo & _other) const
     return (baseType == other.baseType) && (kind == other.kind) && (extra == other.extra);
 }
 
-
 //===========================================================================
 
+template<typename T>
+T* getOrCreateSingletonAtomic(std::atomic<T*>& singleton)
```
Comment (ghalliday) [inline common/deftype/deftype.cpp:1441]: style: I don't like the implicit creation of the object based on the type of the parameter - it is very unclear from the caller what is going to be created.  Better would  be to pass in a factory method similar to querySingleton in jmutex.hpp
note: Is there a reason not to use querySingleton()?

Comment (ghalliday) [inline common/deftype/deftype.cpp:1441]: naming: getLinkedSingleton() would be a better name.  A singleton by definition is either retrieved or created, so no need for orCreate in the function name.  Atomic is also really redundant.

File: common/deftype/deftype.cpp [hunk 3]
```diff
@@ -1623,12 +1651,23 @@ extern DEFTYPE_API ITypeInfo *makeIntType(int len, bool isSigned)
     if (len <= 0 || len > 8)
         return NULL;
 
-    CriticalBlock procedure(*typeCS);
-    CIntTypeInfo *ret = itt[isSigned][len-1];
-    if (ret==NULL)
-        ret = itt[isSigned][len-1] = new CIntTypeInfo(len, isSigned);
-    ::Link(ret);
-    return ret;
+    std::atomic<CIntTypeInfo*>& singleton = itt[isSigned][len-1];
```
Comment (ghalliday) [inline common/deftype/deftype.cpp:1654]: reuse: simpler if this also used querySingleton(), or reused the function above.  (One reason why passing in the function to create the object is a good idea).
Same comment goes for all the other functions for objects with non-default constructors.

File: common/deftype/deftype.cpp [hunk 4]
```diff
@@ -1687,121 +1748,111 @@ extern DEFTYPE_API ITypeInfo *makeBitfieldType(int len, ITypeInfo * basetype)
...
 
 extern DEFTYPE_API ITypeInfo *makeBoolType()
 {
-    CriticalBlock procedure(*typeCS);
-    if (!btt)
-        btt = new CBoolTypeInfo();
-    ::Link(btt);
-    return btt;
+    ITypeInfo* type = getOrCreateSingletonAtomic(btt);
+    return type;
```
Comment (ghalliday) [inline common/deftype/deftype.cpp:1781]: style: why assign to a temporary first?

File: common/deftype/deftype.cpp [hunk 5]
```diff
@@ -1998,6 +2049,16 @@ inline void ReleaseAndClear(T * & ptr)
     }
 }
 
+template <class T>
+inline void ReleaseAndClear(std::atomic<T*>& atomicPtr)
```
Comment (ghalliday) [inline common/deftype/deftype.cpp:2053]: scope: This does not need to be atomic - it is only called single threaded at process termination.

File: common/deftype/deftype.cpp [hunk 6]
```diff
@@ -1623,12 +1651,23 @@ extern DEFTYPE_API ITypeInfo *makeIntType(int len, bool isSigned)
...
-    if (ret==NULL)
-        ret = itt[isSigned][len-1] = new CIntTypeInfo(len, isSigned);
-    ::Link(ret);
-    return ret;
+    std::atomic<CIntTypeInfo*>& singleton = itt[isSigned][len-1];
+    CIntTypeInfo* val = singleton.load();
+    if (val == nullptr)
+    {
+        CIntTypeInfo* newVal = new CIntTypeInfo(len, isSigned);
+        if (!singleton.compare_exchange_strong(val, newVal))
```
Comment (ghalliday) [inline common/deftype/deftype.cpp:1659]: readability: Although this code is correct I had to read it several times to check that val was correctly set up on all the paths.
Add a comment should to the true branch, or change to always assign val to singleton.load() after the if.

**Review by ghalliday — COMMENTED** [review]
Various comments.  I would hope the code would be simpler after this change - it currently is more complicated.

File: common/deftype/deftype.cpp (hunk 2 above)
Comment (jpmcmu) [reply common/deftype/deftype.cpp:1441]: I didn't realize a function like querySingleton already existed I will use that instead.

File: common/deftype/deftype.cpp (hunk 3 above)
Comment (jpmcmu) [reply common/deftype/deftype.cpp:1654]: Agreed, will do.

**Comment by AttilaVamos** [comment]
@jpmcmu I don't think those Regression Test Errors are related to this PR (I saw them today in other PR based on master as well).

**Comment by AttilaVamos** [comment]
@jpmcmu Now it is clean (it was a test system problem).

**Review by ghalliday — COMMENTED** [review]
@jpmcmu looks good - simpler code and improved semantics.  Please squash.

**Review by ghalliday — APPROVED** [review]

**Comment by jpmcmu** [comment]
@ghalliday great, squashed.

**Pull request merged**